src/resource.c

```diff
@@ -625,6 +625,8 @@ coap_add_observer(coap_resource_t *resource,
 
   assert( session );
 
+  /* Fix CID 1484383 */
+  coap_session_reference(session);
```

Comment:
Since `coap_find_observer()` and `coap_find_observer_query()` cannot cause session to be released, I would prefer to just include `coap_delete_observer()` with `coap_session_reference/release`. This would have the advantage that the following three calls to `coap_session_release()` are not necessary, and the code would less complex.

Reply:
Actually, just wrapping `coap_delete_observer()` will not work.  The ref bump must continue to be active until the session is re-referenced just after `669: coap_subscription_init(s);`.

With suggested wrapper:-
Ref = 1 at entry to coap_add_observer().  Bump to 2 just before `coap_delete_observer()`, Drop to 1 inside `coap_delete_observer().` Drop to 0 just after call to `coap_delete_observer()` removes session.

Reply:
Uh. I see...